Added search facets to community & collection pages

alexandrevryghem opened the pull request:
## References
* Fixes #1901
* Related to DSpace/DSpace#9348
* Requires #2722
* Requires #2794

## Description
Added the search section as the default community page tab and moved the community en collection lists to a separate tab. I replaced the recently added section on the collection page with the same search section. I also moved the `comcolSelectionSort` config in the new `comcol` config group.

## Instructions for Reviewers
List of changes in this PR:
- Removed the `CollectionRecentlyAddedComponent`
- Created the new `ComcolSearchSectionComponent`, this is the tab with the search form. It is possible to hide the sidebar facets using the new config `comcol.searchSection.showSidebar`
- Set the `ComcolSearchSectionComponent` section as the default community/collection tab
- Moved the `SubComColSectionComponent` section under the `subcoms-cols` route on community pages (should maybe be renamed depending on community feedback)

Bug fixes:
- Fixed searchevents request sometimes sending an incorrect UUID as the `clickedObject`'s value (it contained the queryParameters)
- The `SearchComponent`'s `trackStatistics` still sends `searchevents` requests when false
- Prevented angular from sending the request `/server/api/submission/vocabularyEntryDetails/search/top?vocabulary=undefined` when the search facet has type `hierarchy` (in `discovery.xml`), but you don't [configure any vocabularies](https://github.com/DSpace/dspace-angular/blob/055122918fb29b33557dada4b8a900eafbe2090e/config/config.example.yml#L380) in you `config.yml` file
- Prevent the access status data from being retrieved when it is not enabled

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

alexandrevryghem commented:
@tdonohue: This PR didn't take much time to create, the big part of the work was already done in #2722 and those changes are compatible with #2275. This PR simply reused the tab from the collection page on the community page. So I think that PR #2275 is indeed a better solution for the problem. I looked over the code and the conflicts actually aren't that big, the only thing that needs to happen is to create a new tab component containing the `<ds-configuration-search-page>` selector.

tdonohue commented:
@alexandrevryghem : Thanks for your feedback.  In that case, I'll review both #2275 and #2722 and leave this PR alone.  We'll resolve the minor conflicts between these other PRs once we merge the first one.  I'm not sure which will be merged first...it will depend on the reviews, but I'll get both reviewed today/tomorrow.

All this means that this PR likely can be closed, as it appears we'll fix this issue via #2275 and #2722.

alexandrevryghem commented:
@tdonohue: I looked at the code of #2275 again and it contains a lot of changes by accident I think, but what it mainly does is:
- Moving the default tab (community & collection list) to a different component in order to change the default tab *(this is already covered by PR #2722)*
- Making the following code the new default tab on community & collection pages *(this is what actually conflicts with this PR)*:
```html
<ds-configuration-search-page [showScopeSelector]="false"></ds-configuration-search-page>
```
- Adding the search component to the home page

Currently PR #2275 doesn't work correctly on communities & collections because the scope isn't given to the `ds-configuration-search-page`, which means that all the items are visible on all the community/collection pages. Last week I implemented this same search functionality for one of our clients and I found that giving the scope to the `SearchComponent` wasn't somthing that was very straightforward. The solution to fix the scope problem should simply be to add a scope `@Input()` to the `SearchComponent` in order to pass it to the `SearchSidebarComponent` & `SearchFormComponent`. This solution does work fine for `SearchFormComponent`, but it doesn't work for `SearchSidebarComponent`, because the filter suggestions & sort options look at the scope in the url (`scope={uuid}`) in order to define in which scope the are and which `discovery.xml` configuration they should use. But because the scope is actually given through the `@Input()` and not the url, the suggestions & searchfilters/sort options think that no scope is defined and therefore the default searchfilters & sort options are shown. Because this was a community bug/improvent I received time to work on this bug this Friday, together with another bug I found while working on that functionality related to the `defaultSortField` field in `discovery.xml`.

The bug described above should be fixed before PR #2275 is merged, so that's why I tought that it might be a better idea to seperate issue DSpace/DSpace#8426 into 2 seperate issues:
- Adding the search to the home page configuration (I would suggest that @GauravD2t would create this PR, since he already wrote the code for that [here](https://github.com/DSpace/dspace-angular/pull/2275/files#diff-706eb8f9b96cb9443381f4484d060637f652f0c587565414b0b19df9e080e87bR5-R6))
- Adding the search to the community & collection page  (Because the scope currently can't be given to the search component and it's children, and that this is required to solve this ticket. I would suggest that I create that bug fix and then I can also add support for search pages with scope to this PR, since it would only require [this change](https://github.com/alexandrevryghem/dspace-angular/commit/db030bd66b3dc160a3c6cf3002d7c11b36d070f3) if I want to add support for that functionality starting from my branch (together with the new `[scope]="collection.id"` attribute that I will create))

tdonohue commented:
@alexandrevryghem : Your suggestion makes sense that we likely need to break down the tasks in https://github.com/DSpace/DSpace/issues/8426 because we have encountered major conflicts between the code in #2275 and the necessary refactors here and in #2722

I did want to point out a small correction:
* #2275 *does* work for community/collection pages.  But, appears to only work because it modifies the behavior of [`collection-page.component.ts`](https://github.com/DSpace/dspace-angular/pull/2275/files#diff-4fe7080da6fffd0f4de5bd360c8f34685893768025f3c1c3c9473e3f8d733a4dR75) and `[community-page.compontent.ts`](https://github.com/DSpace/dspace-angular/pull/2275/files#diff-6ec6184d4eb885bc5bf447f947f98584c10189b12d06df2edf053c7b2ca55c60R80) to hardcode the search `scope` to the `id` of the object being viewed.
    * This approach "works".  However, I agree that it's less ideal than updating the search components themselves to have a `[scope]` parameter.  The latter is better as it allows search components to be widely reused elsewhere without having to hardcode a `scope` in every other component that uses them.
 
I'll reach out to @GauravD2t in PR #2275 and see if we can use the approach you suggested.

alexandrevryghem commented:
@tdonohue: When I tested that PR I used a custom `DiscoveryConfiguration` for a certain scope to verify that you could configure a custom config for certain collections only. But I was able to verify that this issue was already present on DSpace 7.6 so I created issue #2793 and fixed it in a separate PR (#2794) since this might be something worth back porting to 7.6.2.

tdonohue commented:
@alexandrevryghem : When you get a chance, it looks like this PR could now be updated since #2722 is merged.  I suspect that'd also make this much smaller in size / easier to review.

tdonohue commented:
@alexandrevryghem : Friendly reminder to please update/rebase this against latest `main` so that I can review it again.  Our 8.0 feature PR merger deadline is Friday, Feb 23 & I'd love to include this.  Thanks!

alexandrevryghem commented:
@tdonohue: I'll do this now, but just a heads up, this PR will be dependent on #2794

tdonohue commented:
@alexandrevryghem : Oh, thanks for the note!  I didn't realize #2794 was required for this PR.  In that case, I'll try to get #2794 reviewed today/tomorrow to help this along

tdonohue commented on src/app/shared/comcol/sections/comcol-search-section/comcol-search-section.component.ts:
```diff
@@ -0,0 +1,39 @@
+    },
+  ],
+})
+export class ComcolSearchSectionComponent implements OnInit {
```
Can we add a basic description of this component in a comment at the top of the class?  It's just useful for other developers to immediately know where/how it is being used

tdonohue commented on config/config.example.yml:
```diff
@@ -382,10 +382,14 @@ vocabularies:
+  sortConfig:
+    sortField: 'dc.title'
+    sortDirection: 'ASC'
+  searchSection:
```
Please add a comment to describe this setting

tdonohue commented on config/config.example.yml:
```diff
@@ -382,10 +382,14 @@ vocabularies:
-  sortField: 'dc.title'
-  sortDirection: 'ASC'
+# Common community & collection page config
+comcol:
```
I wanted to point out that the old `comcolSelectionSort` config section was **not used** on the Community & Collection pages.  Instead it was used in the Create/Update of new Communities/Collections.
 
I find it a bit odd to lump that setting together with a setting that is specific to Community/Collection pages.  I'm not against it entirely, as I like simplification. But it just needs to be well explained that these configs relate to *both* Communities & Collection but not necessarily just the Community/Collection pages.

The only alternative I can think of would be to leave this `comcolSelectionSort` setting as-is for now & perhaps consider creating your new configs in two places:
```
# This already exists in config.yml
collection;
    searchSection: 
        showSidebar: true

# This does not yet exist, but arguably could.
community:
    searchSection: 
        showSidebar: true
```

That'd allow for more fine-grained control.  But, I admit, I'm not sure if that's necessary for this feature.  So, I won't object to what you've done, but wanted to point out an alternative.

tdonohue reviewed (changes requested):
Thanks @alexandrevryghem !  At a basic level, this works well & the code looks good.  However, I've come across what looks like a few bugs/usability issues.  None of these seem major though.

1. First, when using the search facets on a Community or Collection page, I'm seeing a 422 error on the backend.  It's not visible on the frontend, but appears in the logs & in the Network tab of DevTools. Here's what I'm doing:
    * Install this PR with latest `main` backend
    * Visit a Community or Collection (doesn't matter which)
    * In the Filters, select an Author (doesn't matter who)
    * You'll see this 422 error in logs & in Network tab of DevTools when it attempts to track this search event in statistics via `POST /server/api/statistics/searchevents`
       ```
       Unprocessable or invalid entity (status:422)
          org.dspace.app.rest.exception.UnprocessableEntityException: Error parsing request body
             at org.dspace.app.rest.repository.SearchEventRestRepository.createSearchEvent(SearchEventRestRepository.java:48) ~[dspace-server-webapp-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
             at org.dspace.app.rest.StatisticsRestController.postSearchEvent(StatisticsRestController.java:97) ~[dspace-server-webapp-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
       ```
    * This error is NOT reproducible from the global search page. It only occurs from Community/Collection pages.
2. I find it a bit odd that, by default, on a Community page the first results are all child Collections.  (I have a suggestion of how to fix this in the next bullet)
![all-collections](https://github.com/DSpace/dspace-angular/assets/483997/73c08653-2341-412f-9046-50ee857c02bd)

3. Similar to the previous point.  I find it odd that in both Community/Collection searches, the default is to sort by "Most Relevant".  This results in a very odd ordering of the results (almost random).
    * To solve both this point and the previous one, I'd recommend we default this Sort By to "Accessioned Date Descending".  **That would make the "Search" tab a recently submitted items listing by default.**  It would also ensure Collections do NOT appear first in the list on a Community page.

I also have a few inline comments below.  None of these issues are show stoppers.

alexandrevryghem commented:
@tdonohue: Thnx for the review, I implemented the feedback and wile looking at the network tab I also fixed 2 other small bugs (see last 3 points of the bug fixes section in the PR description). This PR now also requires the PR DSpace/DSpace#9348. While testing I also found out that the sidebar facets didn't work correctly in mobile mode, but this is caused by #1789.

For the searchevents bug, I fixed it so that it wil now always send the `clickedObject` without any query parameters (since this should be a UUID). I did however find it strange that this request was even triggered, since `trackStatistics` is `false` by default. I also added that bug fix to this PR, so if you want to test the bug again you will need to set `trackStatistics` to `true` first
